src/shared/utilities/promiseUtilities.ts

```diff
@@ -0,0 +1,45 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+import * as AsyncLock from 'async-lock'
+
+const lock = new AsyncLock()
+
+/**
+ * Allows callers to use the same promise instead of starting a new one if it is currently running
+ */
+export class PromiseSharer {
+    private static readonly LOCK_PROMISE_REUSE: string = 'lock.promise.reuse'
```

Comment:
For the future: If we re-use `PromiseSharer` in other places, we should find a way to avoid re-using the same lock for unrelated promises. Can you add a TODO to this effect?